Geometry_Engine/Query/ExternalEdges.cs

```diff
@@ -102,38 +108,74 @@ public static List<ICurve> ExternalEdges(this PolySurface surface)
                 return null;
 
-            List<Line> edges = mesh.Faces.SelectMany(f => f.Edges(mesh)).ToList();
+            if (filterByTopology)
+            {
+                List<Tuple<int, int>> allEdgeTopologies = new List<Tuple<int, int>>();
+
+                //Creates tuples of indecies of edges of each face, ordering the index order so the lowest number is always first (to simplify grouping lower down)
+                foreach (Face face in mesh.Faces)
+                {
+                    allEdgeTopologies.Add(new Tuple<int, int>(Math.Min(face.A, face.B), Math.Max(face.A, face.B)));
+                    allEdgeTopologies.Add(new Tuple<int, int>(Math.Min(face.B, face.C), Math.Max(face.B, face.C)));
+                    if (face.D == -1)
+                    {
+                        allEdgeTopologies.Add(new Tuple<int, int>(Math.Min(face.C, face.A), Math.Max(face.C, face.A)));
+                    }
+                    else
+                    {
+                        allEdgeTopologies.Add(new Tuple<int, int>(Math.Min(face.C, face.D), Math.Max(face.C, face.D)));
+                        allEdgeTopologies.Add(new Tuple<int, int>(Math.Min(face.D, face.A), Math.Max(face.D, face.A)));
+                    }
+                }
+
+                IEnumerable<Tuple<int, int>> unique = allEdgeTopologies.GroupBy(x => x).Where(x => x.Count() == 1).SelectMany(x => x);  //Get out instances that are uniqe
 
-            for (int i = edges.Count - 1; i > 0; i--)
+                return unique.Select(x => new Line { Start = mesh.Vertices[x.Item1], End = mesh.Vertices[x.Item2] }).ToList(); //Generate lines based on the unique tuples
+            }
+            else
             {
-                for (int n = i - 1; n >= 0; n--)
+                List<Line> edges = mesh.Faces.SelectMany(f => f.Edges(mesh)).ToList();
+
+                double sqTol = tolerance * tolerance;
+
+                for (int i = edges.Count - 1; i > 0; i--)
                 {
-                    if ((edges[i].Start == edges[n].Start & edges[i].End == edges[n].End) | // edge[i] == edge[n]
-                        (edges[i].Start == edges[n].End & edges[i].End == edges[n].Start))   // edge[i] == edge[n].Reverse()
+                    for (int n = i - 1; n >= 0; n--)
                     {
-                        edges.RemoveAt(i); // shared edge so remove both
-                        edges.RemoveAt(n);
-                        i--;
-                        break;
+                        if ((edges[i].Start.SquareDistance(edges[n].Start) < sqTol && edges[i].End.SquareDistance(edges[n].End) < sqTol) || // edge[i] == edge[n]
+                            (edges[i].Start.SquareDistance(edges[n].End) < sqTol && edges[i].End.SquareDistance(edges[n].Start) < sqTol))   // edge[i] == edge[n].Reverse()
+                        {
+                            edges.RemoveAt(i); // shared edge so remove both
+                            edges.RemoveAt(n);
```

Comment:
Isn't it safer to wait with removal until the end of iteration? Otherwise you may get wrong results in case of nonmanifold topology?

Reply:
Yeah, you are right. Had not changed this really from what it was before, But might as well get it fixed I guess.

Scope for a more deep refactoring of this though, making use of some more cleaver datastructures such as the domain trees. Might have a go at that as well.